fix(plugin-chart-echarts): calculate Gauge Chart intervals correctly when min value is set

goto-loop opened the pull request:
<!---
Please write the PR title following the conventions at https://www.conventionalcommits.org/en/v1.0.0/
Example:
fix(dashboard): load charts correctly
-->

### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

Fixes #21328.

If you create a Gauge Chart with interval bounds and a minimum value set, the color boundaries are placed incorrectly. This PR changes the boundary calculation to take the minimum value into account during normalization by using the formula described [here](https://stats.stackexchange.com/a/70807).

I've also updated the part of the test file where the interval calculation gets tested.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

Before (boundary between green and red should be at value 10 but isn't):
![Screenshot from 2024-02-27 16-58-21](https://github.com/apache/superset/assets/102797966/524860b7-4d41-4ee8-a163-8169f7f2f13a)
After:
![Screenshot from 2024-02-27 16-59-06](https://github.com/apache/superset/assets/102797966/676f341d-048f-45ab-835a-b8eeb3a7d16e)

### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->

1. Create a new Gauge chart, e.g. with exported_stats example dataset
2. Select AVG(daily_members_posting_messages) as Metric
3. On Customize tab, set INTERVAL BOUNDS to 10, 16.53 and INTERVAL COLORS to 3, 6
4. Set MIN to 8
5. Now the boundary between green and red gets updated with the rest of the chart

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [x] Has associated issue: #21328 
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

goto-loop committed:
Take min value into account when normalizing interval bounds

goto-loop committed:
Update test for intervals

villebro reviewed (commented):
I agree with the change, and I was able to validate the change in local testing. However, it would be nice if we could add a few unit tests for `setIntervalBoundsAndColors`, as the `transformProps` test doesn't seem to actually be catching this error.

Also, this is not your making, but calling a function `set` that is in fact not setting, but returning something, feels misleading. I think `getInvervalBoundsAndColors` would be more appropriate.

rusackas commented:
I was hoping to spin up a test environment here, but the docker build is failing. A rebase might fix that if warranted. Thanks @villebro for testing locally.

goto-loop committed:
Change function name to better reflect its purpose

goto-loop committed:
Add unit tests for getIntervalBoundsAndColors

goto-loop commented:
Thanks @villebro for the feedback! I've renamed the function and added some tests specifically for it, please have a look and let me know what you think.

@rusackas I've rebased the branch, maybe you can try the docker build again.

villebro commented:
Thanks @goto-loop for the great tests and addressing the function name! Fingers crossed CI passes 🤞

villebro commented on superset-frontend/plugins/plugin-chart-echarts/test/Gauge/transformProps.test.ts:
```diff
@@ -342,3 +350,43 @@ describe('Echarts Gauge transformProps', () => {
   });
 });
+
+describe('getIntervalBoundsAndColors', () => {
+  it('should generate correct interval bounds and colors', () => {
```
nit/FYI: we're moving away from `describe` with nested `it`s towards simple `test` blocks. But we can convert these in a follow-up refactor pass.

goto-loop replied:
Good point, I wasn't aware of that. Will keep it in mind for the next PR :slightly_smiling_face:

villebro reviewed (approved):
LGTM, thanks for this great fix and quick follow-up on the review comments ❤️
